proxy/src/main/java/com/google/common/util/concurrent/RecyclableRateLimiter.java

```diff
@@ -0,0 +1,138 @@
+package com.google.common.util.concurrent;
+
+import com.google.common.math.LongMath;
+
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+/**
+ * An alternative RateLimiter implementation that allows to "return" unused permits back to the pool to handle retries
```

Comment:
Is it possible to use a normal guava rate limiter, but to continually modify its rate? That would be easier to review at least :), and less error prone.